src/clr-angular/forms/common/form.ts

```diff
@@ -28,4 +49,52 @@ export class ClrForm {
   markAsTouched() {
     this.markControlService.markAsTouched();
   }
+
+  @HostListener('submit')
+  onFormSubmit() {
+    this.markAsTouched();
+    // https://github.com/angular/angular/issues/12463
+    if (isPlatformBrowser(this.platformId)) {
+      const invalidControl = this.el.nativeElement.querySelector('.ng-invalid');
```

Comment:
Why are we using native querying here? Also, we shouldn't rely on Angular's internal class, `.ng-invalid`.
I see that we have a pretty nice structure in our `clrForm`. So can we use Angular querying instead:

```
@ContentChildren(ClrInputContainer) inputContainers: QueryList<ClrInputContainer>;


// and then somewhere in a method

const invalidInputControllers = inputContainers.filter(inputContainer => inputContainer.invalid);

invalidInputControllers[0].setFocusOnInput(); // we can easily add this method in the input container component
```

Reply:
I think the issue with using content children is there would need to be one for each control type (select, radio, input etc). Moving to @ContentChildren introduces some things forms would have to do accounting for:
1. Which controls are invalid - the filter would work but ... there would have to be an @ContentChildren for every form control type (excluding combobox which doesn't have a container yet)
2. How would I know which control is the first invalid control in the Form if they are all separated into different @ContentChildren? .

Reply:
I was suggesting to `@ContentChildren` in `ClrForm` directive. The code snippet I am showing would be used in `ClrForm`. Once you filter only invalid input containers and have them in array, you could do `invalidInputControllers[0]`, which is the first invalid input container. 

And do we even need to concern about the ones without a container? Those ones wouldn't have `clr-control-error`, right?

If you are concerned with inputs not nested in a container, we could directly query `clrInput`, and filter the invalid ones with `descendants: true` flag in `ContentChildren`

Reply:
I don't have a concern with inputs not nested in an clr-input-container - I don't think this behavior should be supported for controls that are not in a container But I also have to account for the other control containers (clr-checkbox-container, clr-toggle-container, clr-date-container, clr-password-container, clr-radio-container, clr-select-container, and clr-textarea-container).  Unless I'm mistaken, it is not possible to have one ContentChildren with multiple Types so forms.ts would end up with multiple @ContentChildren. 

I'm ok with that but would have to modify the logic to find a different way to figure out which control is first so it could be focused. I'm also ok with adding the focus method to the containers but they would all need the change.

Reply:
ok I see what you are saying now. Sorry, I assumed we are `clrInput` in all control container components.

Reply:
Ok I have an idea on to avoid using both querySelector and `.ng-invalid`. Let's discuss that on Slack.